### DISCUSSION TITLE
Referência do usuário autenticado na aplicação

### DISCUSSION THREAD

**Discussion by hosanah** [balta-io/2811#41] [Q&A] [answered]
Fala pessoal, vocês podem me ajudar ? 

No curso de autenticação e autorização, gostaria de saber como faremos pra após o token gerado e enviado para API como podemos pegar a referência do usuário que está logado na aplicação ?

Desde já grato

**Comment by andrebaltieri** [comment, marked as answer]
Boa noite, @hosanah, para obter o usuário logado basta usar o código abaixo em qualquer lugar da Action do Controller...

```csharp
User.Identity.Name
```

**Reply by hosanah** [reply]
Boa Balta, Obrigado!!

`new Claim(ClaimTypes.Name, user.Email),`

Passei aqui, no trecho do código isso, mas isso faz com que, se eu queria pegar o usuário completo do banco, eu tenha que fazer uma consulta no banco pra isso.

dessa forma aqui:

`var user = await context
            .User
            .FirstOrDefaultAsync(x => x.Email == User.Identity.Name);
`

E além de ter que replicar esse código sempre que preciso, seria pertinente utilizar o principio de injeção de dependência pra já deixar esse usuário já dentro da minha aplicação ou to falando besteira Balta ?


Obrigado!

**Reply by andrebaltieri** [reply]
Então, sempre que eu logo um usuário, eu faço isto:

```csharp
var claims = new List<Claim>
  {
      new(ClaimTypes.Name, student.Id.ToString()),
      new(ClaimTypes.GivenName, student.Name),
      new(ClaimTypes.Email, student.Email)
  };
```

Depois eu uso os Extension Methods para adicionar alguns métodos no IdentityUser:
```csharp
public static class IdentityExtension
{
    public static Guid GetId(this ClaimsPrincipal principal)
    {
        try
        {
            return new Guid(principal.Claims.FirstOrDefault(x => x.Type == ClaimTypes.Name)?.Value ?? string.Empty);
        }
        catch
        {
            return Guid.Empty;
        }
    }

    public static string GetName(this ClaimsPrincipal principal)
    {
        try
        {
            return principal.Claims.FirstOrDefault(x => x.Type == ClaimTypes.GivenName)?.Value ?? string.Empty;
        }
        catch
        {
            return string.Empty;
        }
    }

    public static string GetEmail(this ClaimsPrincipal principal)
    {
        try
        {
            return principal.Claims.FirstOrDefault(x => x.Type == ClaimTypes.Email)?.Value ?? string.Empty;
        }
        catch
        {
            return string.Empty;
        }
    }
}
```

Desta forma, eu tenho os método: 
```csharp
User.GetEmail();
User.GetId();
User.GetName();
```

**Reply by hosanah** [reply]
Obrigado Balta!
